portfolio/src/main/java/com/google/sps/servlets/SpotifyServlet.java

```diff
@@ -0,0 +1,150 @@
+    .build();
+
+  private static GetUsersCurrentlyPlayingTrackRequest getUsersCurrentlyPlayingTrackRequest = spotifyApi
+    .getUsersCurrentlyPlayingTrack()
+    .build();
+
+  private static AuthorizationCodeRefreshRequest authorizationCodeRefreshRequest = spotifyApi.authorizationCodeRefresh()
+    .build();
+
+  public static String authorizationCodeUri() {
+    final URI uri = authorizationCodeUriRequest.execute();
+    return uri.toString();
+  }
+
+  public static String formatAsMinutesAndSeconds(Integer milliseconds) {
+    Integer seconds = (Integer)(milliseconds/ 1000) % 60;
+    Integer minutes = (Integer)(milliseconds/ (1000) / 60);
+    return String.format("%d:%02d", minutes, seconds);
+  }
+
+  public static void authorizationCode() {
+    try {
+      final AuthorizationCodeCredentials authorizationCodeCredentials = authorizationCodeRequest.execute();
+      spotifyApi.setAccessToken(authorizationCodeCredentials.getAccessToken());
+      spotifyApi.setRefreshToken(authorizationCodeCredentials.getRefreshToken());
+      System.out.println("Expires in: " + authorizationCodeCredentials.getExpiresIn());
+    } catch (IOException | SpotifyWebApiException | ParseException e) {
+      System.out.println("Error: " + e.getMessage());
+    }
+  }
+
+  public static void authorizationCodeRefresh() {
+    try {
+      final AuthorizationCodeCredentials authorizationCodeCredentials = authorizationCodeRefreshRequest.execute();
+ 
+      spotifyApi.setAccessToken(authorizationCodeCredentials.getAccessToken());
+      spotifyApi.setRefreshToken(authorizationCodeCredentials.getRefreshToken());
+ 
+      System.out.println("Expires in: " + authorizationCodeCredentials.getExpiresIn());
+    } catch (IOException | SpotifyWebApiException | ParseException e) {
+      System.out.println("Error: " + e.getMessage());
+    }
+  }
+
+  public static String getCurrentTrack() {
+    try {
+      final CompletableFuture<CurrentlyPlaying> currentlyPlayingFuture = getUsersCurrentlyPlayingTrackRequest.executeAsync();
+
+      final CurrentlyPlaying currentlyPlaying = currentlyPlayingFuture.join();
+
+      if (currentlyPlaying == null) {
+        return "No song currently playing.";
+      }
+
+      IPlaylistItem item = currentlyPlaying.getItem();
+      String songName = "";
+      String artists = "";
+      
+      if (item instanceof Track) {
+        songName = ((Track)item).getName();
+
+        for (int iterator = 0; iterator < ((Track)item).getArtists().length; iterator++) {
+          if (iterator == ((Track)item).getArtists().length - 1) {
+            artists += ((Track)item).getArtists()[iterator].getName();
+            break;
+          }
+          artists += ((Track)item).getArtists()[iterator].getName() + ", ";
+        }
+      }
+
+      Integer progressMilliseconds = currentlyPlaying.getProgress_ms();
+      Integer durationMilliseconds = ((Track)item).getDurationMs();
+
+      return (songName + " by " + artists + " at " +
```

Comment:
This means if item isn't an instance of Track it'll just display as " by at 1:23 of 4:56" right? Might be worth handling that better

Reply:
I'll make sure to return that string within the if statement, and return a different string if the item isn't an instance of Track.